Enable stumble simulation

crankycoder opened the pull request:
This adds preliminary hooks to enable simulated stumbles.

Overview:

There's a new developer option checkbox to enable simulation mode.

When this is enabled, toggling the stumbler to 'on' will generate a context wrapper (SimulationContext) around the application context that is used by the GPSScanner, WifiScanner and CellScanner.

The SimulationContext provides hooks to emit CellInfo and ScanResult to spoof cell scans and wifi scans.  The GPS scans are spoofed using a MOCK_LOCATION permission and we inject new locations into the LocationManager.

Currently, the spoofed locations will start from your last known 'real' GPS location and will simply increment north west by roughly 5 m every second.  Cell tower and wifi spoofing is hardcoded to simply use a single tower and 2 wifi BSSIDs.

The next step to getting this into something more useful will be to land #1255 and then have the ability to read back the JSON logs and replay a full stumble.

I've renamed a few of the Context instances so that it's clear that we're actually passing the main application context. This is important as that context is wrapped by SimulationContext in the case of a simulation.

The AsyncUploader also checks the simulation preference to see if uploads should be real or just be mocked out with a spoofed HTTP 200 success.

Along the way, I've upgraded robolectric to use revision 2.4.

crankycoder committed:
Added a boolean pref to denote that stumble simulation has been enabled.

crankycoder committed:
Changed the GPSListener to just request the LocationManager from the MapFragment instead of calling through a chain of objects.
Corrected the ClientPrefs type assignment in MainApp.

crankycoder committed:
Remove dead code.

crankycoder committed:
Wrap the context used by the GPSScanner and WifiScanner with a ContextWrapper in the case that we are using a simulated stumble.

crankycoder committed:
Comments on where to add code for #1191

crankycoder committed:
* Added a WifiManagerProxy so that we can properly mock out the
  WifiManager callbacks that provide events for WifiScanner
* rebased back onto dev

crankycoder committed:
added a boolean developer option to enable simulation mode

crankycoder committed:
added hooks so that we can mock out the WifiManager using a WifiManagerProxy

crankycoder committed:
Extracted setup methods for crash preference, simulation preference and map resolution preference.
Wired up the checkbox for simulation so that it persists the setting into ClientPrefs

crankycoder committed:
Added ability to save lat/lon for the start of a simulation run

crankycoder committed:
Set the default start position based on the current GPS lat/lon being received on startup

crankycoder committed:
import cleanups

crankycoder committed:
bumped build to use robolectric 2.4

crankycoder committed:
Cleanup of imports and removed stale comments.

crankycoder committed:
Relaxed the ClientPrefs::getInstance() method so that it doesn't check the type at runtime.  In practice, an invalid cast will cause a similar error for the developer.

crankycoder committed:
* Extracted the CellScannerImpl interface to top level interface SimpleCellScanner
* Renamed CellScannerImplementation to SimpleCellScannerImplementation
* dropped sTestingModeCellInfoArray from CellScanner as it's no longer used

crankycoder committed:
Disable uploads of reports and mock out a HTTP 200 response if simulated stumbles are enabled.

crankycoder committed:
Moved all preferences for simulated stumbling down into Prefs from ClientPrefs

crankycoder committed:
Added a MockSimpleCellScanner
Updated all references to ClientPrefs::isSimulateStumble to use Prefs::isSimulateStumble

crankycoder committed:
Added permission to generate mock locations.

crankycoder committed:
* Added hooks to generate mock CellInfo and ScanInfo data to mock out wifi and cell scans during simulation mode to the ContextWrapper

crankycoder committed:
Updated DeveloperActivity to only use Prefs when writing to the simulation preference.

crankycoder committed:
* Renamed mContext to mAppContext as it's important to be able to track the lifecycle of a context unless it's the application context.
* Changed ScanManager so that mGPSScanner, mWifiScanner and mCellScanner are set to NULL when scanning is stopped.

This is required as start/stop may occur after the simulation flag has changed.

* ScanManager will now wrap the application context with  SimulateStumbleContextWrapper which has all the state information required to run a simulation.

The wrapped context is handed down to the GPSScanner, WifiScanner and CellScanner

crankycoder committed:
* WifiScanner and WifiManagerProxy have been updated to check the simulation preference and inject simulated wifi scans into the intent broadcast system.
* Some renames were done on the contexts to indicate that the expected context is the application context (or wrapped context)

crankycoder committed:
* CellScanner no longer tries to obtain context.getApplicationContext() and instead just hands the existing appContext to the SimpleCellScanner implementation.
* MockSimpleCellScanner hooks the context to grab the next block of fake cell data now.

crankycoder committed:
Merge branch 'features/1191-simulator' into rebased-1191

Conflicts:
	android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/ScanManager.java
	android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/WifiScanner.java
	android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/cellscanner/CellScanner.java
	android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/cellscanner/SimpleCellScannerImplementation.java

crankycoder committed:
Merge branch 'dev' into features/1191-merged

Conflicts:
	android/src/main/java/org/mozilla/mozstumbler/client/subactivities/DeveloperActivity.java
	android/src/main/res/layout/fragment_developer_options.xml

crankycoder committed:
Renamed context wrapper to SimulationContext to save finger typing.

crankycoder committed:
* AndroidManifest.xml will only include a permission request for
  MOCK_LOCATION on debug builds now.
* DeveloperActivity will disable the developer checkbox unless this is
  a debug build.

crankycoder committed:
Merge branch 'dev' into features/1191-merged

Conflicts:
	android/build.gradle

crankycoder commented:
This needs to properly mock out the visible GPS on the GUI and we need to catch an exception if the phone does not have the developer option of "Allow Mock Locations" is enabled.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/client/mapview/GPSListener.java:
```diff
@@ -69,8 +70,14 @@ void removeListener() {
         if (mMapActivity != null) {
             mMapActivity.get().setUserPositionAt(location);
+            ClientPrefs.getInstance().setSimulationLat((float) location.getLatitude());
+            ClientPrefs.getInstance().setSimulationLon((float) location.getLongitude());
```
Do we have to do this on every location? Isn't this just to seed it one time? Writing to the prefs heavily is something I would like to avoid. 
Is there a Build.Debug flag we can wrap this code in?

crankycoder replied:
I'll change this to expose the lat/long in the developer options and have it set once.  If the default is overwritten once, I'll skip the preference write.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/Prefs.java:
```diff
@@ -26,6 +26,12 @@
+    private static final String SIMULATION_LAT = "simulate_lat";
+    private static final String SIMULATION_LON = "simulate_lon";
+
+
```
nit extra CR

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/Prefs.java:
```diff
@@ -159,6 +165,18 @@ protected void setBoolPref(String key, Boolean state) {
+        return getPrefs().getFloat(key, def);
+    }
+
+
```
nit extra CR

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/ScanManager.java:
```diff
@@ -93,24 +99,35 @@ public synchronized void setPassiveMode(boolean on) {
-            mWifiScanner = new WifiScanner(context);
-            mCellScanner = new CellScanner(context);
+
+        Prefs prefs = Prefs.getInstance();
```
Would love this whole block wrapping in Build.Debug or something

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/ScanManager.java:
```diff
@@ -93,24 +99,35 @@ public synchronized void setPassiveMode(boolean on) {
+        mCellScanner = new CellScanner(mAppContext);
+
+        if (prefs != null) {
+            Log.d(LOG_TAG, "Simulation is set to: " + prefs.isSimulateStumble());
```
this could move to after line 121?

crankycoder replied:
Ah, this was just stuffed in during debugging of the feature. I'm just removing this log line as it provides no value to us.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/ScanManager.java:
```diff
@@ -128,6 +145,12 @@ public synchronized boolean stopScanning() {
             return false;
         }
 
+        if (mAppContext instanceof SimulationContext) {
```
I like this check, if we can do more wraps like this, it separates the flow of test code from non-test code. And prevents any performance nits.

crankycoder replied:
This check is mandatory actually - the simulation expects that the app context has been wrapped.  The context acts as a central source of all simulated data.  Checking the type was the only way I knew to safely get at that data.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/WifiManagerProxy.java:
```diff
@@ -0,0 +1,121 @@
+           return getWifiManager().getScanResults();
+       }
+
+
```
oh yes, my meaningful review comment: extra CRs

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/WifiManagerProxy.java:
```diff
@@ -0,0 +1,121 @@
+        return (WifiManager) mAppContext.getSystemService(Context.WIFI_SERVICE);
+    }
+
+
```
extra CR. see aren't I helpful?

crankycoder replied:
dang it.  I'm sure there's a way to tell Intellij to do this automatically when reformatting code.. but TOO MANY DAMN OPTIONS.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/cellscanner/CellScanner.java:
```diff
@@ -34,31 +38,26 @@
-        mCellScannerImplementation = new CellScannerImplementation(context);
+    public CellScanner(Context appCtx) {
+        mAppContext = appCtx;
+        if ((Prefs.getInstance() != null) && Prefs.getInstance().isSimulateStumble()) {
```
we don't normal check if Prefs are null, it is an assumed precondition that the app created this on startup. You could do a Prefs.createGlobalInstance(appCtx) here if you are concerned. Also, this code could be wrapped in a build-time check for debugging build.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/cellscanner/ISimpleCellScanner.java:
```diff
@@ -0,0 +1,14 @@
+
+/**
+ * Created by victorng on 14-11-10.
+ */
```
Remove this, and add an MPL

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/cellscanner/MockSimpleCellScanner.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Created by victorng on 14-11-11.
+ */
+public class MockSimpleCellScanner implements ISimpleCellScanner {
```
remove, add MPL on file

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/cellscanner/SimpleCellScannerImplementation.java:
```diff
@@ -92,6 +92,21 @@ public synchronized void start() {
+
+            if (mPhoneType != TelephonyManager.PHONE_TYPE_GSM
+                && mPhoneType != TelephonyManager.PHONE_TYPE_CDMA) {
+                throw new UnsupportedOperationException("Unexpected Phone Type: " + mPhoneType);
```
Were we always throwing in this condition?

crankycoder replied:
Yeah. I'm not sure why github is saying this bit is new.  The phone type comparison was done way way back in:

https://github.com/garvankeeley/MozStumbler/blob/723f7727507c872d4ff3459866dbc56de8547126/src/org/mozilla/mozstumbler/service/scanners/cellscanner/CellScannerNoWCDMA.java#L73

There's really only 4 phone types anyway, (NONE, GSM, CDMA and SIP).  I'll modify this so that we treat NONE and SIP the same, and drop the if block on 106-109.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/uploadthread/AsyncUploader.java:
```diff
@@ -124,8 +128,19 @@ private void uploadReports(AsyncUploadParam param) {
+                if (prefs != null && prefs.isSimulateStumble()) {
+
+                   result = new HTTPResponse(200,
+                                                       new HashMap<String, List<String>>(),
```
fix indentation

garvankeeley commented:
Other than my nits, and a few questions, this is working great for me on my device

crankycoder committed:
optimized imports and reformatted Prefs.java

crankycoder committed:
Added a preference to see if we've already saved a new start position instead of using the default start for simulations.
Wrapped writing of new default start position with a check to ClientPrefs::isDefaultSimulationLatLong() to write only once.

crankycoder committed:
Added a button to the developer options to reset the default start position of a simulation.

crankycoder committed:
White space effects to WifiManagerProxy

crankycoder committed:
* Added an isDebug check before trying to use the MockSimpleCellScanner when simulation mode is enabled.
* Whitespace nit fixes.

crankycoder committed:
* Wrapped the creation of the simulation context with a check for AppGlobals.isDebug
* Removed some noisey log messages.

crankycoder committed:
MPL fixes

crankycoder committed:
Simplified the check for PHONE_TYPE in SimpleCellScannerImplementation.

crankycoder commented:
Crap. 3-way merge left to fix up, but I think I've addressed the comments you made.

crankycoder committed:
Merge branch 'dev' into features/1191-merged

Conflicts:
	android/build.gradle
	android/src/main/java/org/mozilla/mozstumbler/client/ClientPrefs.java
	android/src/main/java/org/mozilla/mozstumbler/client/subactivities/DeveloperActivity.java
	android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/ScanManager.java
	android/src/main/res/layout/fragment_developer_options.xml
	android/src/main/res/values/strings.xml

crankycoder commented:
I want to double check the merge conflicts one more time to make sure I got the low power battery mode stuff merged properly.

garvankeeley commented:
+1 to the isDebug changes to ensure these code paths are debug build only. Yeah this looks ready to land.

crankycoder commented:
I've double checked the 3-way merge, this should be ok now.

crankycoder committed:
just a comment and a suggestion so that we don't clobber each other on
3-way merges in dev options anymore.

crankycoder commented:
I'm finishing up adding a testcase to this.  Closing in the meantime.
